File: 프로그래머스/4/1843. 사칙연산/사칙연산.cpp
#include <vector>
#include <string>
#include <algorithm>
#define MX 205
using namespace std;

int solution(vector<string> arr)
{   
    int dp[2], idx = arr.size() - 1, sum = 0;
    
    while(idx >= 0){
        if(arr[idx][0] >= '0'){
            sum += stoi(arr[idx]);
        } else if(arr[idx] == "-") {
            int tmp[2];
            tmp[0] = dp[0], tmp[1] = dp[1];
            dp[0] = max(sum + tmp[0] - 2 * (stoi(arr[idx+1])), -(sum + tmp[1]));
            dp[1] = min(-sum + tmp[1], -(sum + tmp[0]));
            sum = 0;
        }
        idx--;
    }
   
    return sum + dp[0];
}

File: 프로그래머스/4/1843. 사칙연산/사칙연산.java
import java.util.*;

class Solution {
    public int solution(String arr[]) {
        int n = arr.length / 2 + 1;
        int[][] mn = new int[n][n];
        int[][] mx = new int[n][n];
        
        for(int i = 0; i < n; i++) {
            mn[i][i] = Integer.parseInt(arr[i * 2]);
            mx[i][i] = Integer.parseInt(arr[i * 2]);
        }
        
        for(int len = 1; len < n; len++){
            for(int i = 0; i < n - len; i++) {
                int j = i + len;
                mx[i][j] = Integer.MIN_VALUE;
                mn[i][j] = Integer.MAX_VALUE;
                
                for(int k = i; k < j; k++) {
                    String op = arr[k * 2 + 1];
                   
                    int mxLeft = mx[i][k];
                    int mnLeft = mn[i][k];
                    int mxRight = mx[k + 1][j];
                    int mnRight = mn[k + 1][j];
                    
                    if(op.equals("-")) {
                        mx[i][j] = Math.max(mx[i][j], mxLeft - mnRight);
                        mn[i][j] = Math.min(mn[i][j], mnLeft - mxRight);
                    } else {
                        mx[i][j] = Math.max(mx[i][j], mxLeft + mxRight);
                        mn[i][j] = Math.min(mn[i][j], mnLeft + mnRight);
                    } 
                }
            }
        }
        
        return mx[0][n - 1];
    }
}